substrate-node/pallets/pallet-tfgrid/src/grid_migration.rs

```diff
@@ -243,15 +243,32 @@ pub fn migrate_farms<T: Config>() -> frame_support::weights::Weight {
             }
         }
 
-        let replaced_farm_name = farm::replace_farm_name_spaces_with_underscores(&farm.name);
-        let name = match <T as Config>::FarmName::try_from(replaced_farm_name) {
+        let mut farm_name = farm.name;
+        let truncated = farm_name.len() > <T as Config>::MaxFarmNameLength::get() as usize;
+        if truncated {
+            info!("farm name length exceeded, truncating farm name");
+            farm_name.truncate(<T as Config>::MaxFarmNameLength::get() as usize);
+        }
+
+        let replaced_farm_name = farm::replace_farm_name_invalid_characters(&farm_name);
+        if replaced_farm_name != farm_name || truncated {
+            info!("farm name changed, removing from farm id by name map");
+            FarmIdByName::<T>::remove(&farm_name);
+        }
+
+        let name = match <T as Config>::FarmName::try_from(replaced_farm_name.clone()) {
             Ok(n) => n,
             Err(_) => {
                 info!("invalid farm name, skipping updating farm {:?} ...", k);
                 return None;
             }
         };
 
+        if replaced_farm_name != farm_name || truncated {
```

Comment:
This is problematic. If a name despite transformation is not conform the requirements, you effectively removed it , and leave a dangling old farm which can (presumably) no longer be decoded. This means that at this point, the farm is effectively destroyed, which can not happen. You also can't assume that "currently all farms pass" since there might be a farm which gets created just before the upgrade, which then violates the rule and causes mayhem during the upgrade.

Reply:
So what do we do to solve this? If a farm name is invalid, it shouldn't have been created in the first place

Reply:
The fact is that it is there. A solution could be to generate a completely random name, verify it does not exist, then assign that. This is a bit extreme, but as long as a farmer can change his farm name (not sure if that is possible yet but it should, as the name is technically just metadata), that should work

Reply:
Farmer can change his name yes